pkg/gds/admin/v2/csurf.go

```diff
@@ -0,0 +1,100 @@
+package admin
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/rs/zerolog/log"
+)
+
+// Parameters and names for double-cookie submit CSRF protection
+const (
+	CSRFCookie          = "csrf_token"
+	CSRFReferenceCookie = "csrf_reference_token"
+	CSRFHeader          = "X-CSRF-TOKEN"
+)
+
+// DoubleCookie is Cross Site Request Forgery (CSRF/XSRF) protection middleware that
+// checks the presence of an X-CSRF-TOKEN header containing a cryptographically random
+// token that matches a token contained in the CSRF-TOKEN cookie in the request.
+// Because of the same-origin policy, an attacker cannot access the cookies or scripts
+// of the safe site therefore it will not be able to guess what token to put in the
+// request, thereby protecting from CSRF attacks. Note that Double Cookie CSRF
+// protection requires TLS to prevent MITM attacks.
+func DoubleCookie() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		cookie, err := c.Cookie(CSRFReferenceCookie)
+		if err != nil {
+			log.Warn().Err(err).Msg("no csrf token cookie in request")
+			c.JSON(http.StatusForbidden, ErrorResponse(ErrCSRFVerification))
+			c.Abort()
+			return
+		}
+
+		header := c.GetHeader(CSRFHeader)
+		if header, err = url.QueryUnescape(header); err != nil {
```

Comment:
Apologizes in advances for the somewhat lengthy answer, but I wanted to make sure I captured everything to make sure I didn't make a mistake in the code. 

In order to protect URLs so that they can be properly parsed, certain characters are reserved, e.g. `/`, `:`, `.`, `&`, `=`, `?`, whitespace, etc. URL Query Escaping is a somewhat unique way of ensuring a URL is parsed correctly when these characters need to be used in a URL (usually in the query params, e.g. the part that follows the question mark in  `https://example.com/path/to/document.html?key=foo&otherkey=bar`). For historical reasons, unsafe characters are replaced by a "%" and two hexadecimal digits. 

Cookies need to be URL safe*, so many frameworks and browsers URL escape their values. In our case, the token is a base64 encoded string which means it contains `a-z`, `A-Z`, and `0-9`; however that's only 62 characters, two additional characters are needed, which are `+` and `/` in a standard base64 implementation** (which need to be escaped). Base64 encoding also uses a padding character***, `=` if the byte array is not a multiple of 3 bytes, and this character also needs to be escaped. 

Basically what that means is that if we have a token that looks like this: `a-base64-encoded_string=`; it's possible it will be received by the server as `a%2Fbase64%2Fencoded%2Bstring%3D`. The `%` encoding needs to be converted back into unescaped characters in order for us to perform our token comparison otherwise it will fail. 

`*` - this is only partially true and actually not true in our case, but that doesn't stop browsers or HTTP clients from URL escaping cookies; I noticed it using the Go http client in the test cases.

`**` - we're using a URL-safe base64 encoding which uses `_` and `-` as the extra characters which don't need to be escaped; however the `=` padding character is still involved. 

`***` - technically there is something called `RawURLEncoding` that omits padding, but we need a cryptographic equality on our bytes array.